IPython/html/static/widgets/js/widget.js

```diff
@@ -136,12 +136,48 @@ define(["widgets/js/manager",
              * Handle incoming comm msg.
              */
             var method = msg.content.data.method;
+            
             var that = this;
             switch (method) {
                 case 'update':
                     this.state_change = this.state_change
                         .then(function() {
-                            return that.set_state(msg.content.data.state);
+                            var state = msg.content.data.state || {};
+                            var buffer_keys = msg.content.data.buffers || [];
+                            var buffers = msg.buffers || [];
+                            var metadata = msg.content.data.metadata || {};
+                            var i,k;
+                            for (var i=0; i<buffer_keys.length; i++) {
+                                k = buffer_keys[i];
+                                state[k] = buffers[i];
+                            }
+
+                            // for any metadata specifying a deserializer, set the
+                            // state to a promise that resolves to the deserialized version
+                            // also, store the serialization function for the attribute
+                            var keys = Object.keys(metadata);
+                            
+                            for (var i=0; i<keys.length; i++) {
+                                var k = keys[i];
+                                if (metadata[k] && metadata[k].serialization) {
+                                    var deserialize = utils.load_class.call(metadata[k].serialization)
+                                    that.serialize[k] = deserialize;
+                                    state[k] = deserialize.then((function(input) {
```

Comment:
This is a little misleading.  When I first read it I didn't understand what was happening until I got down to the send sync msg method.  I suggest s/`deserialize`/`serializer` and s/`that.serialize`/`that.serializer`.  My thinking is that a `serializer` may contain both or either `serialize` and `deserialize` methods.